addons/auth_passkey/models/res_users.py

```diff
@@ -23,3 +30,45 @@ def action_create_passkey(self):
                 'dialog_size': 'medium',
             }
         }
+
+    @classmethod
+    def _login(cls, db, login, credential, user_agent_env):
+        if credential['type'] == 'webauthn':
+            webauthn = json.loads(credential['content'])
+            with registry(db).cursor() as cr:
+                identifier = base64.urlsafe_b64decode(webauthn['id'] + '===').hex()
+                cr.execute("""
+                    SELECT login
+                      FROM auth_passkey_key key
+                      JOIN res_users usr ON usr.id = key.create_uid
+                     WHERE credential_identifier=%s
+                """, (identifier,))
+                res = cr.fetchone()
+                if not res:
+                    raise AccessDenied(_('Unknown passkey'))
+                login = res[0]
+        return super()._login(db, login, credential, user_agent_env=user_agent_env)
+
+    def _check_credentials(self, credential, env):
+        if credential['type'] == 'webauthn':
+            webauthn = json.loads(credential['content'])
+            identifier = base64.urlsafe_b64decode(webauthn['id'] + '===').hex()
+            passkey = self.env['auth.passkey.key'].sudo().search([
+                ("create_uid", "=", self.env.user.id),
+                ("credential_identifier", "=", identifier),
+            ])
+            if not passkey:
+                raise AccessDenied(_('Unknown passkey'))
+            try:
+                new_sign_count = self.env['auth.passkey.key']._verify_auth(
+                    webauthn,
+                    request.session.pop('webauthn_challenge'),
+                    passkey.public_key,
+                    passkey.sign_count,
+                )
+            except InvalidAuthenticationResponse as e:
+                raise AccessDenied(e.args[0])
+            passkey.sign_count = new_sign_count
+            request.session['skip_totp'] = True
```

Comment:
I am bit concerned with this.
If the code crashes between this moment and the moment of the `session.pop('skip_totp')` applies, then the `skip_totp` remains in the session, and an attacker could leverage this to bypass the 2FA during the next login attempt. For instance it uses the login/password login with this `skip_totp` left over from the previous crashed authentication.

Tough I do not have a miracle solution either.

Out of the blue now I was thinking maybe to store this on the env or the env transaction. The env / transaction is unique by request, and hence by db and user, and is created at the beginning of the transaction and then garbage collected after the request, so there is no possibility to make that `skip_totp` survive longer than one HTTP request, while storing it on the session there is a possibility, even if unlikely, but still a possibility.

But, in `session.authenticate`, a completely new cursor and environment is used:
```py
with registry.cursor() as cr:
    env = odoo.api.Environment(cr, pre_uid, {})
```
so the env/transaction wouldn't be the same as in here in `_check_credentials`. So if you set a variable, let's stay `env.transaction.require_mfa = False` in `_check_credentials`, you wouldn't find it back in that `session.authenticate`.

Another possibility I see is to change the return signature of `res.users.authenticate` to give in the returned value if the MFA should be asked, which you can make True/False according to the authentication used.
Coming all back from the override of `_check_credentials`.

odoo/addons/base/models/res_users.py
```py
def authenticate(cls, db, credential, user_agent_env):
    ...
    return uid, False
```

addons/auth_totp/models/res_users.py
```py
def authenticate(cls, db, credential, user_agent_env):
    ...
    uid, require_mfa = super().authenticate(db, credential, user_agent_env)
    ...
    return uid, require_mfa or env['res.users'].browse(uid).totp_enabled
```

addons/auth_passkey/models/res_users.py
```py
def authenticate(cls, db, credential, user_agent_env):
    ...
    uid, require_mfa = super().authenticate(db, credential, user_agent_env)
    ...
    return uid, False if credential.get('webauthn_reponse') or require_mfa
```
Or delay that to `_check_credentials`, it has the advantage we already have a browse record on user and we could decide to require_mfa not only on the kind of authentication used but also making it depend on data on the user if we want.

addons/auth_totp/models/res_users.py
```py
def _check_credentials(self, credential, env):
    ...
    require_mfa = super()._check_credentials(credential, env)
    ...
    return require_mfa or self.totp_enabled
```

addons/auth_passkey/models/res_users.py
```py
def _check_credentials(self, credential, env):
        if credential['webauthn']:
            ...
            require_mfa = False
            return require_mfa
        else:
            return super()._check_credentials(credential, env)
```

Or another possibility, we change the signature of `_mfa_type` to pass the credential dict or at least the type of authentication:

odoo/addons/base/models/res_users.py
```py
def _mfa_type(self, credential):
        """ If an MFA method is enabled, returns its type as a string. """
        return
```

addons/auth_totp/models/res_users.py
```py
def _mfa_type(self, credential):
        r = super()._mfa_type()
        if r is not None:
            return r
        if self.totp_enabled:
            return 'totp'
```

addons/auth_passkey/models/res_users.py
```py
def _mfa_type(self, credential):
        if credential.get('webauthn_response'):
            return None
        else:
            return super()._mfa_type()
```

The disadvantage of that second alternative with `_mfa_type` in my opinion is that an attacker could try to leverage to make a confused credential dict, e.g. `{'login': 'foo', 'password': 'foo', 'webauthn_response': '...'}`,
then we do not actually know what authentication was really used, is it the login/password ? Is it webauthn ? It entirely depend on the trust of the code implementation, that if the credential dict holds the `webauthn` value, it's definitely went through the `webauthn` authentication and nothing else. So it's less future proof and less custom module / override proof. While if we do the `authenticate`/`_check_credentials` alternative, we can ensure 100% to set `require_mfa` to `False` only when it's the webauthn authenticate which was used.

Though I do not want to take such a decision alone.
I really need the opinion of @xmo-odoo and @odony on this one.